src/PactNet.AspNetCore/ProviderState/ProviderStateMiddleware.cs

```diff
@@ -0,0 +1,133 @@
+using Newtonsoft.Json;
+using PactNet.Verifier.ProviderState;
+
+namespace PactNet.AspNetCore.ProviderState
+{
+    /// <summary>
+    /// Defines the provider state middleware
+    /// It provides an endpoint to trigger the provider states before verification of pact interaction
+    /// </summary>
+    public class ProviderStateMiddleware
+    {
+        /// <summary>
+        /// The next pointer
+        /// </summary>
+        private readonly RequestDelegate next;
+        
+        /// <summary>
+        /// The provider state middleware options
+        /// </summary>
+        private readonly ProviderStateOptions options;
+
+        /// <summary>
+        /// The provider state accessor
+        /// </summary>
+        private readonly IProviderStateAccessor providerStateAccessor;
+
+        ///// <summary>
+        ///// the after interaction callback
+        ///// </summary>
+        //private Dictionary<string, Action> afterInteractionCallbacks;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ProviderStateMiddleware"/>
+        /// </summary>
+        /// <param name="options">the middleware options</param>
+        /// <param name="providerStateAccessor">the provider state accessor</param>
+        /// <param name="next">the next handle</param>
+        public ProviderStateMiddleware(IOptions<ProviderStateOptions> options, IProviderStateAccessor providerStateAccessor, RequestDelegate next)
+        {
+            this.options = options.Value;
+            this.providerStateAccessor = providerStateAccessor;
+            this.next = next;
+        }
+
+        public async Task InvokeAsync(HttpContext context)
+        {
+            if (!(context.Request.Path.Value?.StartsWith(options.RouteProviderState) ?? false))
+            {
+                await this.next(context);
+                return;
+            }
+
+            if (context.Request.Method != HttpMethod.Post.ToString())
+            {
+                await WriteErrorInResponseAsync(context, "The provider state invocation needs to be a post method. Check on your verifier side if the provider state url is nicely configured.", HttpStatusCode.BadRequest);
+                return;
+            }
+
+            var providerStateBody = await ReadProviderStateInteractionAsync(context);
+            var stateHandler = this.providerStateAccessor.GetByDescriptionAndAction(providerStateBody.State,
+                    providerStateBody.Action);
+
+            if (stateHandler != null)
+            {
+                ExecuteState(providerStateBody, stateHandler);
+            }
+            context.Response.StatusCode = (int)HttpStatusCode.OK;
+
+            await WriteToResponseAsync(context, string.Empty);
+        }
+
+        private static void ExecuteState(ProviderStateInteraction providerStateBody, IStateHandler stateHandler)
+        {
+            if (providerStateBody.Params == null || !providerStateBody.Params.Any())
```

Comment:
The design here seems a little unsafe to me - if you define a setup which takes arguments but the invocation doesn't provide any then it's ambiguous what should happen:

- Should it invoke the action anyway, even with no arguments? But then what happens to the templated parts?
- Should it error instead? I'd vote yes, given the invocation is incompatible with the defined provider state

To secure that, I'd have two different interfaces for `IStateHandler` - one which takes args and one which doesn't. That way you can't even attempt to invoke a handler which requires args from an invocation that didn't provide any (or vice versa).

Reply:
Hey Adam, I'm finally having some bandwidth to work on finishing up this PR

I'm having now 2 interfaces implementing IStateHandler with different execute method. 

On the middleware, I'm having something like this : 
`        /// <summary>
        /// Execute a provider state
        /// </summary>
        /// <param name="providerStateBody">The provider state from the consumer contract</param>
        /// <param name="stateHandler">The state handler from the provider</param>
        private static void ExecuteState(ProviderStateInteraction providerStateBody, IStateHandler stateHandler)
        {
            switch (stateHandler)
            {
                case IStateHandlerWithArgs handler when providerStateBody.HasParameters():
                    handler.Execute(providerStateBody.Params);
                    break;
                case IStateHandlerWithoutArgs handlerWithoutArgs:
                    handlerWithoutArgs.Execute();
                    break;
                default:
                    throw new InvalidOperationException(
                        "The provider state in the contract is not compatible with the state definition. Please re-configure your state handler correctly.");
            }
        }`
        
        Does that sound like what you were mentioning when talking interface segregation here?